### PR TITLE
BUGFIX: Prevent refetching nodes for policy checks

### LINKED ISSUE #2301: Policy evaluation re-fetching nodes
### Description

Policies are evaluated in the `/neos/ui-services/get-policy-info` for every node and policy.

When declaring a policy that includes another node (such as `matcher: 'isDescendantNodeOf("93c96eac-4f66-41ce-a8e1-77482f5a5475")'`) the fixed node (in this case `93c96eac-4f66-41ce-a8e1-77482f5a5475`) will be fetched from the database for every subject node that is being compared.

This results in heavy performance hits for all websites that are medium sized or larger. Depending on the webserver setup (e.g. nginx timing out  after some time) the request may not even go through correctly resulting in editors not being able to edit content.

I have narrowed this down to `NodePrivilegeContext::getNodeByIdentifier` where the node is fetched from the database - explicitly resetting the upstream cache since this request is being done without authorization checks.

Adding a small cache (e.g. as static property of `NodePrivilegeContext`) resolves this issue.

In my development environment response times for the AJAX Endpoint dropped from around 200s to about 8s.

### Steps to Reproduce

1. Define a policy including `isDescendantNodeOf(...)`
2. Add a logging statement to `NodePrivilegeContext::getNodeByIdentifier`
3. Open Backend and wait for AJAX request to `/neos/ui-services/get-policy-info`
4. Watch logging output fetch the same node over and over.

#### Expected behavior

Nodes used in policies should be fetched from the database once.

#### Actual behavior

Nodes used in policies are fetched from the database for every subject (basically every node of the page).

### Affected Versions

Neos: At least 3.3.16 and higher. Have not tested lower.

### PR DESCRIPTION
**Pull request by j6s**
When defining a policy targeting a static node the static node was
refetched for every subject, slowing down policy information retrieval.

This adds a simple caching mechanism to `NodePrivilegeContext::getNodeByIdentifier`
that prevents nodes from being refetched for static policy comparisons.

Resolves #2301

### CODE REVIEW

**Comment by kitsunet** [comment]
Everything before 3.3 LTS is no longer bugfixed. Only actual security problems will be fixed in there, so I moved this up to the 3.3 branch.

**Comment by kitsunet** [comment]
Actually caching this as static class member is not a good idea. It's hell to test and can easily create side effects. I would much rather see this as an instance variable (maybe of a shared cache object)

**Comment by aertmann** [comment]
I agree with @kitsunet here, this kind of caching doesn't belong here.

The context factory caches contexts https://github.com/neos/neos-development-collection/blob/master/Neos.ContentRepository/Classes/Domain/Service/ContextFactory.php#L90-L96 and each context caches it's nodes https://github.com/neos/neos-development-collection/blob/master/Neos.ContentRepository/Classes/Domain/Service/Context.php#L269-L280

Additional each node factory caches nodes through it's node factory https://github.com/neos/neos-development-collection/blob/master/Neos.ContentRepository/Classes/Domain/Factory/NodeFactory.php#L81-L89

However the culprit here is that the node privilege removes the node from the context's cache in `$context->getFirstLevelNodeCache()->setByIdentifier($nodeIdentifier, null);`, not sure why this is done though.

**Comment by j6s** [comment]
Without looking any further I would assume that this is done because the method explicitly skips authorization checks. If you would leave nodes in a cache that is accesible through other means this could result in this method inserting nodes into a cache that should not be there in the first place (since authorization checks would prevent that).

**Comment by aertmann** [comment]
Actually I see why, it's to avoid other parts from accessing the node afterwards without checking for access. In this case something like a cache for nodes fetched without security checks might make sense, so it explicitly has to be disabled to access those nodes.

File: Neos.ContentRepository/Classes/Service/Utility/UnvalidatedNodeCache.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ */
+class UnvalidatedNodeCache
+{
+    private $nodes = [];
+
+    /**
+     * @param string $nodeIdentifier
+     * @return bool
+     */
+    public function has($nodeIdentifier)
```
Comment (kdambekalns) [inline Neos.ContentRepository/Classes/Service/Utility/UnvalidatedNodeCache.php]:
```suggestion
    private function has($nodeIdentifier)
```

File: Neos.ContentRepository/Classes/Service/Utility/UnvalidatedNodeCache.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        return array_key_exists($nodeIdentifier, $this->nodes);
+    }
+
+    /**
+     * @param string $nodeIdentifier
+     * @return NodeInterface
+     */
+    public function get($nodeIdentifier)
```
Comment (kdambekalns) [inline Neos.ContentRepository/Classes/Service/Utility/UnvalidatedNodeCache.php]:
```suggestion
    private function get($nodeIdentifier)
```

File: Neos.ContentRepository/Classes/Service/Utility/UnvalidatedNodeCache.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        return $this->nodes[$nodeIdentifier];
+    }
+
+    /**
+     * @param string $nodeIdentifier
+     * @param NodeInterface $node
+     */
+    public function put($nodeIdentifier, NodeInterface $node)
```
Comment (kdambekalns) [inline Neos.ContentRepository/Classes/Service/Utility/UnvalidatedNodeCache.php]:
```suggestion
    private function put($nodeIdentifier, NodeInterface $node)
```

File: Neos.ContentRepository/Classes/Service/Utility/UnvalidatedNodeCache.php [hunk 4]
```diff
@@ -0,0 +1,62 @@
+<?php
+namespace Neos\ContentRepository\Service\Utility;
+
+use Neos\ContentRepository\Domain\Model\NodeInterface;
+use Neos\Flow\Annotations as Flow;
+
+/**
+ * Simple cache for nodes that is intended to be used in `NodePrivilegeContext`.
+ * WARNING: This cache may contain nodes that should not be visible to the current user.
```
Comment (kdambekalns) [inline Neos.ContentRepository/Classes/Service/Utility/UnvalidatedNodeCache.php]:
```suggestion
  *
  * WARNING: This cache may contain nodes that should not be visible to the current user.
```

File: Neos.ContentRepository/Classes/Service/Utility/UnvalidatedNodeCache.php (hunk 1 above)
Comment (j6s) [reply Neos.ContentRepository/Classes/Service/Utility/UnvalidatedNodeCache.php]: ✅ fixed

File: Neos.ContentRepository/Classes/Service/Utility/UnvalidatedNodeCache.php (hunk 3 above)
Comment (j6s) [reply Neos.ContentRepository/Classes/Service/Utility/UnvalidatedNodeCache.php]: ✅ fixed

File: Neos.ContentRepository/Classes/Service/Utility/UnvalidatedNodeCache.php (hunk 4 above)
Comment (j6s) [reply Neos.ContentRepository/Classes/Service/Utility/UnvalidatedNodeCache.php]: ✅ fixed

File: Neos.ContentRepository/Classes/Service/Utility/UnvalidatedNodeCache.php (hunk 2 above)
Comment (j6s) [reply Neos.ContentRepository/Classes/Service/Utility/UnvalidatedNodeCache.php]: ✅ fixed

File: Neos.ContentRepository/Classes/Service/Utility/TransientNodeCache.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
+<?php
+namespace Neos\ContentRepository\Service\Utility;
+
+use Neos\ContentRepository\Domain\Model\NodeInterface;
+use Neos\Flow\Annotations as Flow;
+
+/**
+ * Simple cache for nodes that is intended to be used in `NodePrivilegeContext`.
+ *
+ * WARNING: This cache may contain nodes that should not be visible to the current user.
```
Comment (daniellienert) [inline Neos.ContentRepository/Classes/Service/Utility/TransientNodeCache.php]: If this class should only be used in NodePrivilegeContext I would suggest to move it somewhere int that folder structure.

File: Neos.ContentRepository/Classes/Service/Utility/TransientNodeCache.php [hunk 6]
```diff
@@ -0,0 +1,63 @@
+<?php
```
Comment (daniellienert) [inline Neos.ContentRepository/Classes/Service/Utility/TransientNodeCache.php]:
```suggestion
<?php
declare(strict_types=1);

```

File: Neos.ContentRepository/Classes/Service/Utility/TransientNodeCache.php [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+ */
+class TransientNodeCache
+{
+    private $nodes = [];
+
+    /**
+     * @param string $nodeIdentifier
+     * @return bool
+     */
+    private function has($nodeIdentifier)
```
Comment (daniellienert) [inline Neos.ContentRepository/Classes/Service/Utility/TransientNodeCache.php]:
```suggestion
    private function has(string $nodeIdentifier): bool
```

File: Neos.ContentRepository/Classes/Service/Utility/TransientNodeCache.php [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    {
+        return array_key_exists($nodeIdentifier, $this->nodes);
+    }
+
+    /**
+     * @param string $nodeIdentifier
+     * @return NodeInterface
+     */
+    private function get($nodeIdentifier)
```
Comment (daniellienert) [inline Neos.ContentRepository/Classes/Service/Utility/TransientNodeCache.php]:
```suggestion
    private function get(string $nodeIdentifier): NodeInterface
```

File: Neos.ContentRepository/Classes/Service/Utility/TransientNodeCache.php [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    {
+        return $this->nodes[$nodeIdentifier];
+    }
+
+    /**
+     * @param string $nodeIdentifier
+     * @param NodeInterface $node
+     */
+    private function put($nodeIdentifier, NodeInterface $node)
```
Comment (daniellienert) [inline Neos.ContentRepository/Classes/Service/Utility/TransientNodeCache.php]:
```suggestion
    private function put(string $nodeIdentifier, NodeInterface $node)
```

File: Neos.ContentRepository/Classes/Service/Utility/TransientNodeCache.php [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    {
+        $this->nodes[$nodeIdentifier] = $node;
+    }
+
+    /**
+     * @param string $nodeIdentifier
+     * @param callable $getter
+     * @return NodeInterface
+     */
+    public function cache($nodeIdentifier, callable $getter)
```
Comment (daniellienert) [inline Neos.ContentRepository/Classes/Service/Utility/TransientNodeCache.php]:
```suggestion
    public function cache(string $nodeIdentifier, callable $getter): NodeInterface
```

File: Neos.ContentRepository/Classes/Service/Utility/TransientNodeCache.php (hunk 6 above)
Comment (kdambekalns) [reply Neos.ContentRepository/Classes/Service/Utility/TransientNodeCache.php]: Well, it's to be introduced as of recently… https://discuss.neos.io/t/usage-of-php-strict-typing-by-default/3861

**Comment by kdambekalns** [comment]
I agree to the type declarations, did not add them because I was confused as to support in PHP 7.0 (since this goes into 3.3). But since it doesn't use `void`, that should be fine!

File: Neos.ContentRepository/Classes/Service/Utility/TransientNodeCache.php (hunk 6 above)
Comment (kdambekalns) [reply Neos.ContentRepository/Classes/Service/Utility/TransientNodeCache.php]: Actually it should have a newline more…
```suggestion
<?php

declare(strict_types=1);
```

**Comment by daniellienert** [comment]
> .. I was confused as to support in PHP 7.0 ...
Also looked it up (again), but return types are not as new as one might think :)

File: Neos.ContentRepository/Classes/Security/Authorization/Privilege/Node/TransientNodeCache.php [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        $this->nodes[$nodeIdentifier] = $node;
+    }
+
+    /**
+     * @param string $nodeIdentifier
+     * @param callable $getter
+     * @return NodeInterface
+     */
+    public function cache(string $nodeIdentifier, callable $getter)
```
Comment (daniellienert) [inline Neos.ContentRepository/Classes/Security/Authorization/Privilege/Node/TransientNodeCache.php]: Return type is missing here

File: Neos.ContentRepository/Classes/Security/Authorization/Privilege/Node/TransientNodeCache.php [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+ */
+class TransientNodeCache
+{
+    private $nodes = [];
+
+    /**
+     * @param string $nodeIdentifier
+     * @return bool
+     */
+    private function has(string $nodeIdentifier): bool
```
Comment (daniellienert) [inline Neos.ContentRepository/Classes/Security/Authorization/Privilege/Node/TransientNodeCache.php]: I would remove this private methods. IMHO it is a lot of code and indirection to wrap some simpel methods.

File: Neos.ContentRepository/Classes/Security/Authorization/Privilege/Node/TransientNodeCache.php [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+ * Simple cache for nodes that is intended to be used in `NodePrivilegeContext`.
+ *
+ * WARNING: This cache may contain nodes that should not be visible to the current user.
+ *          Only use this cache in `NodePrivilegeContext`.
+ *
+ * @Flow\Scope("singleton")
+ */
+class TransientNodeCache
+{
+    private $nodes = [];
```
Comment (daniellienert) [inline Neos.ContentRepository/Classes/Security/Authorization/Privilege/Node/TransientNodeCache.php]:
```suggestion
    /**
    * @var NodeInterface[]
    */
    private $nodes = [];
```

**Comment by daniellienert** [comment]
@kdambekalns: Looks good for me now - would you have another look?

**Comment by albe** [comment]
Oh my, we already had two approvals. Now this needs to be retargeted to 4.3 - can we get a quick assertion this is still good to go? I would take care of the rest then @daniellienert @kitsunet @kdambekalns

File: Neos.ContentRepository/Classes/Security/Authorization/Privilege/Node/TransientNodeCache.php [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+     * @return NodeInterface
+     */
+    public function cache(string $nodeIdentifier, callable $getter): NodeInterface
+    {
+        if (array_key_exists($nodeIdentifier, $this->nodes)) {
+            return $this->nodes[$nodeIdentifier];
+        }
+
+        $node = $getter();
+        if ($node) {
```
Comment (daniellienert) [inline Neos.ContentRepository/Classes/Security/Authorization/Privilege/Node/TransientNodeCache.php:38]:
```suggestion
        if ($node instance of NodeInterface) {
```
Nitpick, check pricisely

**Pull request merged**